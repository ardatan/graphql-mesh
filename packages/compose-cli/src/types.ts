import { DocumentNode, GraphQLSchema } from 'graphql';
import { Logger } from '@graphql-mesh/types';
import { fetch as defaultFetch } from '@whatwg-node/fetch';

export interface MeshComposeCLIConfig {
  /**
   * The output destination of the resulting composed GraphQL schema.
   * By default, the CLI will write the result to stdout.
   */
  output?: string;
  subgraphs: MeshComposeCLISubgraphConfig[];
  transforms?: MeshComposeCLITransformConfig[];
  additionalTypeDefs?: string | DocumentNode | (string | DocumentNode)[];
  fetch?: typeof defaultFetch;
  cwd?: string;
}

export interface MeshComposeCLISubgraphConfig {
  sourceHandler: MeshComposeCLISourceHandlerDef;
  transforms?: MeshComposeCLITransformConfig[];
}

export type MeshComposeCLISourceHandlerDef = (ctx: LoaderContext) => {
  name: string;
  schema$: Promise<GraphQLSchema> | GraphQLSchema;
};

export type MeshComposeCLITransformConfig = (input: GraphQLSchema, ...args: any[]) => GraphQLSchema;

export interface LoaderContext {
  fetch: typeof defaultFetch;
  cwd: string;
  logger: Logger;
}

/**
 * Type helper for defining the config.
 */
export function defineConfig(config: MeshComposeCLIConfig) {
  return config;
}
