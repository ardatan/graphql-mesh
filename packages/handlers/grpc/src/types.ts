import {
  GraphQLType,
  GraphQLObjectType,
  GraphQLInt,
  GraphQLFloat,
  GraphQLString,
  GraphQLBoolean,
} from 'graphql';

import {
  BigIntResolver as GraphQLBigInt,
} from 'graphql-scalars';

export interface GrpcGraphqlSchemaConfiguration {
  endpoint: string;
  protoFilePath: string;
  serviceName: string;
  packageName: string;
}

export interface TypeMapping {
  [key: string]: GraphQLType;
}

// https://developers.google.com/protocol-buffers/docs/proto3#scalar
// https://www.apollographql.com/docs/apollo-server/schemas/types.html
export const GRPC_GQL_TYPE_MAPPING: TypeMapping = {
  int32: GraphQLInt,
  int64: GraphQLBigInt, // TODO: https://github.com/graphql/graphql-js/issues/292
  float: GraphQLFloat,
  double: GraphQLFloat,
  string: GraphQLString,
  bool: GraphQLBoolean,
};

export const typeDefinitionCache: TypeMapping = {
  ServerStatus: new GraphQLObjectType({
    name: 'ServerStatus',
    description: 'status of the server',
    fields: () => ({
      status: {
        type: GraphQLString,
        descripton: 'status string',
      },
    }),
  }),
};
