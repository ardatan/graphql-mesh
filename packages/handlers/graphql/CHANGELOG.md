# @graphql-mesh/graphql

## 0.14.1

### Patch Changes

- Updated dependencies [191a663a]
  - @graphql-mesh/types@0.35.1

## 0.14.0

### Minor Changes

- b9ca0c30: Make Transforms and Handlers base-dir aware

### Patch Changes

- Updated dependencies [b9ca0c30]
  - @graphql-mesh/types@0.35.0
  - @graphql-mesh/utils@0.9.0

## 0.13.19

### Patch Changes

- Updated dependencies [ec89a923]
  - @graphql-mesh/utils@0.8.8

## 0.13.18

### Patch Changes

- Updated dependencies [55327fd6]
  - @graphql-mesh/types@0.34.1

## 0.13.17

### Patch Changes

- Updated dependencies [76051dd7]
  - @graphql-mesh/types@0.34.0

## 0.13.16

### Patch Changes

- Updated dependencies [646d6bdb]
  - @graphql-mesh/types@0.33.0

## 0.13.15

### Patch Changes

- Updated dependencies [68d6b117]
  - @graphql-mesh/types@0.32.0

## 0.13.14

### Patch Changes

- Updated dependencies [212f2d66]
  - @graphql-mesh/types@0.31.1

## 0.13.13

### Patch Changes

- Updated dependencies [77327988]
  - @graphql-mesh/types@0.31.0

## 0.13.12

### Patch Changes

- Updated dependencies [48f38a4a]
  - @graphql-mesh/types@0.30.1

## 0.13.11

### Patch Changes

- Updated dependencies [938cca26]
  - @graphql-mesh/types@0.30.0

## 0.13.10

### Patch Changes

- Updated dependencies [8ef29de1]
  - @graphql-mesh/types@0.29.4

## 0.13.9

### Patch Changes

- Updated dependencies [a02d86c3]
- Updated dependencies [a02d86c3]
- Updated dependencies [a02d86c3]
  - @graphql-mesh/types@0.29.3

## 0.13.8

### Patch Changes

- Updated dependencies [69d2198d]
  - @graphql-mesh/utils@0.8.7

## 0.13.7

### Patch Changes

- Updated dependencies [8e8848e1]
  - @graphql-mesh/types@0.29.2

## 0.13.6

### Patch Changes

- Updated dependencies [7e970f09]
  - @graphql-mesh/utils@0.8.6

## 0.13.5

### Patch Changes

- Updated dependencies [e8994875]
  - @graphql-mesh/types@0.29.1

## 0.13.4

### Patch Changes

- Updated dependencies [8d345721]
  - @graphql-mesh/utils@0.8.5

## 0.13.3

### Patch Changes

- Updated dependencies [c767df01]
- Updated dependencies [183cfa96]
- Updated dependencies [b3d7ecbf]
  - @graphql-mesh/types@0.29.0
  - @graphql-mesh/utils@0.8.4

## 0.13.2

### Patch Changes

- Updated dependencies [a22fc6f3]
  - @graphql-mesh/types@0.28.0

## 0.13.1

### Patch Changes

- Updated dependencies [c1de3e43]
  - @graphql-mesh/types@0.27.0

## 0.13.0

### Minor Changes

- 75f6dff9: feat(graphql): ability to disable batch execution

### Patch Changes

- Updated dependencies [75f6dff9]
- Updated dependencies [c4f207a7]
  - @graphql-mesh/types@0.26.0

## 0.12.1

### Patch Changes

- d4ed4e89: fix(graphql): schemaHeaders bug

## 0.12.0

### Minor Changes

- 0df817d0: feat(graphql): support exported schemaHeaders

### Patch Changes

- Updated dependencies [0df817d0]
  - @graphql-mesh/types@0.25.0

## 0.11.3

### Patch Changes

- Updated dependencies [08c2966e]
  - @graphql-mesh/utils@0.8.3

## 0.11.2

### Patch Changes

- Updated dependencies [b6262481]
  - @graphql-mesh/types@0.24.0

## 0.11.1

### Patch Changes

- Updated dependencies [e5b38574]
  - @graphql-mesh/types@0.23.3

## 0.11.0

### Minor Changes

- 7140ef09: feat(graphql): support custom variables in endpoint url

## 0.10.5

### Patch Changes

- Updated dependencies [c85a54eb]
  - @graphql-mesh/utils@0.8.2

## 0.10.4

### Patch Changes

- Updated dependencies [c614e796]
  - @graphql-mesh/types@0.23.2

## 0.10.3

### Patch Changes

- Updated dependencies [59d77fb8]
  - @graphql-mesh/types@0.23.1

## 0.10.2

### Patch Changes

- 44e14ca5: fix(graphql): handle null variables in multipart requests

## 0.10.1

### Patch Changes

- Updated dependencies [e5cd44f5]
  - @graphql-mesh/types@0.23.0

## 0.10.0

### Minor Changes

- 2fd59a83: feat(graphql): add useSSEForSubscription option to use SSE for Subscriptions instead of WebSocket

### Patch Changes

- Updated dependencies [2fd59a83]
  - @graphql-mesh/types@0.22.0

## 0.9.2

### Patch Changes

- 86f359cb: fix(graphql): fix handling urls

## 0.9.1

### Patch Changes

- Updated dependencies [c064e3a8]
  - @graphql-mesh/types@0.21.1
  - @graphql-mesh/utils@0.8.1

## 0.9.0

### Minor Changes

- 03f41cd0: feat(graphql): support exported GraphQLSchema from code files

### Patch Changes

- 1f0b2f1f: fix(graphql): fix multipart issues
- Updated dependencies [03f41cd0]
  - @graphql-mesh/types@0.21.0

## 0.8.1

### Patch Changes

- 1e7fd602: feat(graphql): add `multipart` option to support file uploads
- Updated dependencies [1e7fd602]
  - @graphql-mesh/types@0.20.1

## 0.8.0

### Minor Changes

- 2d14fcc3: bump graphql-tools to v7
- 2d14fcc3: feat(graphql): use graphql-ws instead of subscriptions-transport-ws

### Patch Changes

- Updated dependencies [2d14fcc3]
- Updated dependencies [2d14fcc3]
  - @graphql-mesh/types@0.20.0
  - @graphql-mesh/utils@0.8.0

## 0.7.10

### Patch Changes

- Updated dependencies [c1b073de]
  - @graphql-mesh/types@0.19.0
  - @graphql-mesh/utils@0.7.0

## 0.7.9

### Patch Changes

- Updated dependencies [5628fb14]
  - @graphql-mesh/types@0.18.0

## 0.7.8

### Patch Changes

- Updated dependencies [0560e806]
  - @graphql-mesh/types@0.17.1

## 0.7.7

### Patch Changes

- Updated dependencies [c26c8c56]
  - @graphql-mesh/types@0.17.0

## 0.7.6

### Patch Changes

- Updated dependencies [3770af72]
  - @graphql-mesh/types@0.16.1

## 0.7.5

### Patch Changes

- Updated dependencies [3ee10180]
  - @graphql-mesh/types@0.16.0

## 0.7.4

### Patch Changes

- Updated dependencies [0f17c58c]
  - @graphql-mesh/types@0.15.0

## 0.7.3

### Patch Changes

- Updated dependencies [937c87d2]
  - @graphql-mesh/types@0.14.1

## 0.7.2

### Patch Changes

- Updated dependencies [1e0445ee]
  - @graphql-mesh/types@0.14.0

## 0.7.1

### Patch Changes

- Updated dependencies [b50a68e3]
  - @graphql-mesh/types@0.13.0

## 0.7.0

### Minor Changes

- e2b34219: improve introspection caching

### Patch Changes

- Updated dependencies [e2b34219]
- Updated dependencies [9a7a55c4]
  - @graphql-mesh/types@0.12.0
  - @graphql-mesh/utils@0.6.0

## 0.6.4

### Patch Changes

- a3b42cfd: fix(runtime): handle transforms correctly for single source
- Updated dependencies [2dedda3c]
- Updated dependencies [a3b42cfd]
  - @graphql-mesh/types@0.11.3
  - @graphql-mesh/utils@0.5.4

## 0.6.3

### Patch Changes

- Updated dependencies [6d624576]
  - @graphql-mesh/types@0.11.2

## 0.6.2

### Patch Changes

- Updated dependencies [405cec23]
  - @graphql-mesh/types@0.11.1

## 0.6.1

### Patch Changes

- Updated dependencies [48d89de2]
  - @graphql-mesh/types@0.11.0

## 0.6.0

### Minor Changes

- 79adf4b6: feat(config): support functions as config property

### Patch Changes

- Updated dependencies [79adf4b6]
- Updated dependencies [79adf4b6]
  - @graphql-mesh/utils@0.5.3
  - @graphql-mesh/types@0.10.0

## 0.5.5

### Patch Changes

- Updated dependencies [2d5cc25b]
  - @graphql-mesh/types@0.9.2

## 0.5.4

### Patch Changes

- Updated dependencies [93ad5255]
  - @graphql-mesh/types@0.9.1

## 0.5.3

### Patch Changes

- Updated dependencies [9900d2fa]
- Updated dependencies [9900d2fa]
  - @graphql-mesh/utils@0.5.2

## 0.5.2

### Patch Changes

- Updated dependencies [c8d9695e]
- Updated dependencies [8f53be10]
  - @graphql-mesh/types@0.9.0
  - @graphql-mesh/utils@0.5.1

## 0.5.1

### Patch Changes

- Updated dependencies [d2e56567]
  - @graphql-mesh/types@0.8.1

## 0.5.0

### Minor Changes

- a789c312: feat(stitching): use batch execution

### Patch Changes

- Updated dependencies [a789c312]
  - @graphql-mesh/types@0.8.0
  - @graphql-mesh/utils@0.5.0

## 0.4.0

### Minor Changes

- 718e7a16: fix(runtime): fix browser support

### Patch Changes

- Updated dependencies [718e7a16]
  - @graphql-mesh/types@0.7.0
  - @graphql-mesh/utils@0.4.0

## 0.3.2

### Patch Changes

- Updated dependencies [5067ac73]
- Updated dependencies [a76d74bb]
  - @graphql-mesh/types@0.6.0

## 0.3.1

### Patch Changes

- Updated dependencies [dde7878b]
  - @graphql-mesh/types@0.5.1

## 0.3.0

### Minor Changes

- 705c4626: introduce an independent config package

### Patch Changes

- Updated dependencies [705c4626]
  - @graphql-mesh/types@0.5.0
  - @graphql-mesh/utils@0.3.0

## 0.2.19

### Patch Changes

- Updated dependencies [854dc550]
- Updated dependencies [6f21094b]
  - @graphql-mesh/types@0.4.0

## 0.2.18

### Patch Changes

- Updated dependencies [3c131332]
  - @graphql-mesh/types@0.3.1

## 0.2.17

### Patch Changes

- Updated dependencies [ccede377]
  - @graphql-mesh/types@0.3.0

## 0.2.16

### Patch Changes

- 1d32f9bf: Fix handling GraphQL variables with useGETForQueries
