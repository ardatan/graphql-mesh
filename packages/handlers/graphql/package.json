{
  "name": "@graphql-mesh/graphql",
  "version": "0.14.1",
  "sideEffects": false,
  "main": "dist/index.cjs.js",
  "module": "dist/index.esm.js",
  "typings": "dist/index.d.ts",
  "typescript": {
    "definition": "dist/index.d.ts"
  },
  "license": "MIT",
  "peerDependencies": {
    "graphql": "*"
  },
  "dependencies": {
    "@graphql-mesh/types": "0.35.1",
    "@graphql-mesh/utils": "0.9.0",
    "fetchache": "0.0.4",
    "@graphql-tools/url-loader": "6.8.2",
    "@graphql-tools/wrap": "7.0.5"
  },
  "publishConfig": {
    "access": "public",
    "directory": "dist"
  }
}
