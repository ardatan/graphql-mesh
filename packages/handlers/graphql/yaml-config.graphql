extend type Handler {
  """
  Handler for remote/local/third-party GraphQL schema
  """
  graphql: GraphQLHandler
}

type GraphQLHandler @md {
  """
  JSON object representing the Headers to add to the runtime of the API calls
  """
  headers: JSON
  """
  A url to your remote GraphQL endpoint
  """
  endpoint: String!
}
