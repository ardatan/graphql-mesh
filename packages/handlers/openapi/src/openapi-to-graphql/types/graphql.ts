// Copyright IBM Corp. 2018. All Rights Reserved.
// Node module: openapi-to-graphql
// This file is licensed under the MIT License.
// License text available at https://opensource.org/licenses/MIT

/**
 * Custom type definitions for GraphQL.
 */

import {
  GraphQLObjectType,
  GraphQLScalarType,
  GraphQLInputObjectType,
  GraphQLList,
  GraphQLEnumType,
  GraphQLUnionType,
  GraphQLResolveInfo,
} from 'graphql';

export enum GraphQLOperationType {
  Query,
  Mutation,
  // TODO: Subscription
}

export type GraphQLType =
  | GraphQLObjectType
  | GraphQLInputObjectType
  | GraphQLList<any>
  | GraphQLUnionType
  | GraphQLEnumType
  | GraphQLScalarType;

type Arg = {
  type: any;
  description?: string;
};

export type Args = {
  [key: string]: Arg;
};

export type ResolveFunction = (root: any, args: any, ctx: any, info: GraphQLResolveInfo) => Promise<any> | any;

export type Field = {
  type: GraphQLType;
  resolve?: ResolveFunction;
  args?: Args;
  description: string;
};
