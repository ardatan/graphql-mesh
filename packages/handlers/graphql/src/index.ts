import { GetMeshSourceOptions, MeshHandler, MeshSource, ResolverData, YamlConfig } from '@graphql-mesh/types';
import { fetchache, KeyValueCache, Request } from 'fetchache';
import { UrlLoader } from '@graphql-tools/url-loader';
import { GraphQLSchema, buildClientSchema, introspectionFromSchema, buildSchema } from 'graphql';
import { introspectSchema } from '@graphql-tools/wrap';
import {
  getInterpolatedHeadersFactory,
  ResolverDataBasedFactory,
  getHeadersObject,
  loadFromModuleExportExpression,
  getInterpolatedStringFactory,
} from '@graphql-mesh/utils';
import { ExecutionParams, AsyncExecutor } from '@graphql-tools/delegate';

export default class GraphQLHandler implements MeshHandler {
  private name: string;
  private config: YamlConfig.GraphQLHandler;
  private baseDir: string;
  private cache: KeyValueCache<any>;

  constructor({ name, config, baseDir, cache }: GetMeshSourceOptions<YamlConfig.GraphQLHandler>) {
    this.name = name;
    this.config = config;
    this.baseDir = baseDir;
    this.cache = cache;
  }

  async getMeshSource(): Promise<MeshSource> {
    const { endpoint, schemaHeaders: configHeaders, cacheIntrospection, introspection } = this.config;

    if (endpoint.endsWith('.js') || endpoint.endsWith('.ts')) {
      const schema = await loadFromModuleExportExpression<GraphQLSchema>(endpoint, { cwd: this.baseDir });
      return {
        schema,
      };
    } else if (endpoint.endsWith('.graphql')) {
      const rawSDL = await loadFromModuleExportExpression<string>(endpoint, { cwd: this.baseDir });
      const schema = buildSchema(rawSDL);
      return {
        schema,
      };
    }
    const urlLoader = new UrlLoader();
    const customFetch: WindowOrWorkerGlobalScope['fetch'] = (...args) =>
      fetchache(args[0] instanceof Request ? args[0] : new Request(...args), this.cache);
    const getExecutorAndSubscriberForParams = (
      params: ExecutionParams,
      headersFactory: ResolverDataBasedFactory<Headers>,
      endpointFactory: ResolverDataBasedFactory<string>
    ) => {
      const resolverData: ResolverData = {
        root: {},
        args: params.variables,
        context: params.context,
      };
      const headers = getHeadersObject(headersFactory(resolverData));
      const endpoint = endpointFactory(resolverData);
      return urlLoader.getExecutorAndSubscriberAsync(endpoint, {
        customFetch: customFetch as any,
        ...this.config,
        headers,
      });
    };
    let schema: GraphQLSchema;
    let schemaHeaders =
      typeof configHeaders === 'string'
        ? await loadFromModuleExportExpression(configHeaders, { cwd: this.baseDir })
        : configHeaders;
    if (typeof schemaHeaders === 'function') {
      schemaHeaders = schemaHeaders();
    }
    if (schemaHeaders && 'then' in schemaHeaders) {
      schemaHeaders = await schemaHeaders;
    }
    const schemaHeadersFactory = getInterpolatedHeadersFactory(schemaHeaders || {});
    const introspectionExecutor: AsyncExecutor = async (params): Promise<any> => {
      const { executor } = await getExecutorAndSubscriberForParams(params, schemaHeadersFactory, () => endpoint);
      return executor(params);
    };
    if (introspection) {
      const result = await urlLoader.handleSDLAsync(introspection, {
        customFetch: customFetch as any,
        ...this.config,
        headers: schemaHeaders,
      });
      schema = result.schema;
    } else if (cacheIntrospection) {
      const cacheKey = this.name + '_introspection';
      const introspectionData: any = await this.cache.get(cacheKey);
      if (introspectionData) {
        schema = buildClientSchema(introspectionData);
      } else {
        schema = await introspectSchema(introspectionExecutor);
        const ttl = typeof cacheIntrospection === 'object' && cacheIntrospection.ttl;
        const introspection = introspectionFromSchema(schema);
        this.cache.set(cacheKey, introspection, { ttl });
      }
    } else {
      schema = await introspectSchema(introspectionExecutor);
    }
    const operationHeadersFactory = getInterpolatedHeadersFactory(this.config.operationHeaders);
    const endpointFactory = getInterpolatedStringFactory(endpoint);
    return {
      schema,
      executor: async params => {
        const { executor } = await getExecutorAndSubscriberForParams(params, operationHeadersFactory, endpointFactory);
        return executor(params) as any;
      },
      subscriber: async params => {
        const { subscriber } = await getExecutorAndSubscriberForParams(
          params,
          operationHeadersFactory,
          endpointFactory
        );
        return subscriber(params);
      },
      batch: 'batch' in this.config ? this.config.batch : true,
    };
  }
}
