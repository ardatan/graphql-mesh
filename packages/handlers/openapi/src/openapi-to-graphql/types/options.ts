// Copyright IBM Corp. 2018. All Rights Reserved.
// Node module: openapi-to-graphql
// This file is licensed under the MIT License.
// License text available at https://opensource.org/licenses/MIT

// Type imports:
import { GraphQLOperationType, SubscriptionContext } from './graphql';
import { GraphQLFieldResolver, GraphQLResolveInfo } from 'graphql';
import { ResolverMiddleware } from '../resolver_builder';
import { Logger, MeshPubSub } from '@graphql-mesh/types';

/**
 * Type definition of the options that users can pass to OpenAPI-to-GraphQL.
 */
export type Warning = {
  type: string;
  message: string;
  mitigation: string;
  path?: string[];
};

export type Report = {
  warnings: Warning[];
  numOps: number;
  numOpsQuery: number;
  numOpsMutation: number;
  numOpsSubscription: number;
  numQueriesCreated: number;
  numMutationsCreated: number;
  numSubscriptionsCreated: number;
};

export type ConnectOptions = {
  [key: string]: boolean | number | string;
};

// A standardized type that can be used to identify a specific operation in an OAS
export type OasTitlePathMethodObject<T> = {
  [title: string]: {
    [path: string]: {
      [method: string]: T;
    };
  };
};

export type Headers = { [key: string]: string };

/**
 * Given a set parameters corresponding to a specific operation in the OAS,
 * provide the appropriate headers
 */
export type RequestHeadersFunction<TSource, TContext, TArgs> = (
  method: string,
  path: string,
  title: string,
  resolverParams?: {
    source: TSource;
    args: TArgs;
    context: TContext;
    info: GraphQLResolveInfo;
  }
) => Headers;

/*
export type RequestOptions<TSource, TContext, TArgs> = Omit<
  RequestInit,
  'headers'
> & {
  headers?: Headers | RequestHeadersFunction<TSource, TContext, TArgs>
}
*/
export type RequestOptions<TSource, TContext, TArgs> = RequestInit;
export type Options<TSource, TContext, TArgs> = Partial<InternalOptions<TSource, TContext, TArgs>>;

export type InternalOptions<TSource, TContext, TArgs> = {
  /*
   * Adhere to the OAS as closely as possible. If set to true, any deviation
   * from the OAS will lead OpenAPI-to-GraphQL to throw.
   */
  strict: boolean;

  /**
   * Holds information about the GraphQL schema generation process
   */
  report: Report;

  // Schema options

  /**
   * Field names can only be sanitized operationIds
   *
   * By default, query field names are based on the return type type name and
   * mutation field names are based on the operationId, which may be generated
   * if it does not exist.
   *
   * This option forces OpenAPI-to-GraphQL to only create field names based on the
   * operationId.
   */
  operationIdFieldNames: boolean;

  /**
   * Under certain circumstances (such as response code 204), some RESTful
   * operations should not return any data. However, GraphQL objects must have
   * a data structure. Normally, these operations would be ignored but for the
   * sake of completeness, the following option will give these operations a
   * placeholder data structure. Even though the data structure will not have
   * any practical use, at least the operations will show up in the schema.
   */
  fillEmptyResponses: boolean;

  /**
   * Auto-generate a 'limit' argument for all fields that return lists of
   * objects, including ones produced by links
   *
   * Allows to constrain the return size of lists of objects
   *
   * Returns the first n number of elements in the list
   */
  addLimitArgument: boolean;

  /**
   * If a schema is of type string and has format UUID, it will be translated
   * into a GraphQL ID type. To allow for more customzation, this option allows
   * users to specify other formats that should be interpreted as ID types.
   */
  idFormats?: string[];

  /**
   * Allows to define the root operation type (Query or Mutation type) of any
   * OAS operation explicitly.
   *
   * OpenAPI-to-GraphQL will by default make all GET operations Query fields and all other
   * operations into Mutation fields.
   *
   * The field is identifed first by the title of the OAS, then the path of the
   * operation, and lastly the method of the operation.
   */
  selectQueryOrMutationField?: OasTitlePathMethodObject<GraphQLOperationType>;

  /**
   * Sets argument name for the payload of a mutation to 'requestBody'
   */
  genericPayloadArgName: boolean;

  /**
   * By default, field names are sanitized to conform with GraphQL conventions,
   * i.e. types should be in PascalCase, fields should be in camelCase, and
   * enum values should be in ALL_CAPS.
   *
   * This option will prevent OpenAPI-to-GraphQL from enforcing camelCase field names and
   * PascalCase type names, only removing illegal characters and staying as true
   * to the provided names in the OAS as possible.
   */
  simpleNames: boolean;

  /**
   * Experimental feature that will try to create more meaningful names from
   * the operation path than the response object by leveraging common
   * conventions.
   *
   * For example, given the operation 'GET /users/{userId}/car', OpenAPI-to-GraphQL will
   * create a Query field 'userCar'. Note that because 'users' is followed by
   * the parameter 'userId', it insinuates that this operation will get the car
   * that belongs to a singular user. Hence, the name 'userCar' is more fitting
   * than 'usersCar' so the pluralizing 's' is dropped.
   *
   * This option will also consider irregular plural forms.
   */
  singularNames: boolean;

  // Resolver options

  /**
   * Custom headers to send with every request made by a resolve function.
   */
  headers?: Headers | RequestHeadersFunction<TSource, TContext, TArgs>;

  /**
   * Custom query parameters to send with every reqeust by a resolve function.
   */
  qs?: { [key: string]: string };

  /**
   * Allows to override or add options to the node's request object used to make
   * calls to the API backend.
   * e.g. Setup the web proxy to use.
   */
  requestOptions?: RequestOptions<TSource, TContext, TArgs>;

  /**
   * Allows to override or add options to the PubSub connect object used to make
   * publish/subscribe to the API backend.
   * e.g. Setup the web proxy to use.
   */
  connectOptions?: ConnectOptions;

  /**
   * Specifies the URL on which all paths will be based on.
   * Overrides the server object in the OAS.
   */
  baseUrl?: string;

  /**
   * Allows to define custom resolvers for fields on the Query/Mutation root
   * operation type.
   *
   * In other words, instead of resolving on an operation (REST call) defined in
   * the OAS, the field will resolve on the custom resolver. Note that this will
   * also affect the behavior of links.
   *
   * The field is identifed first by the title of the OAS, then the path of the
   * operation, and lastly the method of the operation.
   *
   * Use cases include the resolution of complex relationships between types,
   * implementing performance improvements like caching, or dealing with
   * non-standard authentication requirements.
   */
  customResolvers?: OasTitlePathMethodObject<GraphQLFieldResolver<TSource, TContext, TArgs>>;

  /**
   * Allows to define custom resolvers and subscribe functions for fields on the
   * Subscription root operation type.
   *
   * In other words, instead of resolving on an operation (REST call) defined in
   * the OAS, the field will resolve on the custom resolver. Note that this will
   * also affect the behavior of links.
   *
   * The field is identifed first by the title of the OAS, then the path of the
   * operation, and lastly the method of the operation.
   *
   * Use cases include the resolution of complex relationships between types,
   * implementing performance improvements like caching, or dealing with
   * non-standard authentication requirements.
   *
   * Note: Subscription fields will only be generated if the
   * createSubscriptionsFromCallbacks option is enabled.
   */
  customSubscriptionResolvers?: OasTitlePathMethodObject<{
    subscribe: GraphQLFieldResolver<TSource, SubscriptionContext, TArgs>;
    resolve: GraphQLFieldResolver<TSource, TContext, TArgs>;
  }>;

  // Authentication options

  /**
   * Determines whether OpenAPI-to-GraphQL should create viewers that allow users to pass
   * basic auth and API key credentials.
   */
  viewer: boolean;

  /**
   * JSON path to OAuth 2 token contained in GraphQL context. Tokens will per
   * default be sent in "Authorization" header.
   */
  tokenJSONpath?: string;

  /**
   * Determines whether to send OAuth 2 token as query parameter instead of in
   * header.
   */
  sendOAuthTokenInQuery: boolean;

  // Logging options

  /**
   * The error extensions is part of the GraphQLErrors that will be returned if
   * the query cannot be fulfilled. It provides information about the failed
   * REST call(e.g. the method, path, status code, response
   * headers, and response body). It can be useful for debugging but may
   * unintentionally leak information.
   *
   * This option prevents the extensions from being created.
   */
  provideErrorExtensions: boolean;

  /**
   * Appends a small statement to the end of field description that clarifies
   * the operation that the field will trigger.
   *
   * Will affect query and mutation fields as well as fields created from links
   *
   * In the form of: 'Equivalent to {title of OAS} {method in ALL_CAPS} {path}'
   * Will forgo the title is only one OAS is provided
   */
  equivalentToMessages: boolean;

  /**
   * Custom fetch implementation
   */
  fetch: typeof import('cross-fetch').fetch;

  /**
   * Middlewares for Resolver Factory
   */
  resolverMiddleware: ResolverMiddleware<TSource, TContext, TArgs>;

  /**
   * Includes HTTP Details to the result object
   */
  includeHttpDetails?: boolean;
  pubsub: MeshPubSub;
  logger: Logger;
};
