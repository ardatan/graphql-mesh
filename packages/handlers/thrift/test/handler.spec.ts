import thriftHandler from '../src';
import { Hooks } from '@graphql-mesh/types';
import { InMemoryLRUCache } from '@graphql-mesh/cache-inmemory-lru';
import { join } from 'path';
import { printSchema } from 'graphql';
import { EventEmitter } from 'events';

describe('thrift', () => {
  it('should create a GraphQL Schema from Thrift IDL', async () => {
    const source = await thriftHandler.getMeshSource({
      name: 'Twitter',
      config: {
        idl: join(__dirname, './fixtures/twitter.thrift'),
        hostName: 'localhost',
        port: 4444,
        path: '/twitter',
        serviceName: 'twitter-service',
      },
      cache: new InMemoryLRUCache(),
      hooks: new EventEmitter() as Hooks,
    });
    expect(printSchema(source.schema)).toMatchSnapshot();
  });
});
