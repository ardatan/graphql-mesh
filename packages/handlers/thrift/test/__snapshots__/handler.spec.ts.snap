// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`thrift should create a GraphQL Schema from Thrift IDL 1`] = `
"type Query {
  \\"\\"\\"\\"\\"\\"
  ping: Boolean

  \\"\\"\\"\\"\\"\\"
  postTweet(tweet: TweetInput): Boolean

  \\"\\"\\"\\"\\"\\"
  searchTweets(query: String): TweetSearchResult

  \\"\\"\\"\\"\\"\\"
  zip: Boolean
}

\\"\\"\\"\\"\\"\\"
input TweetInput {
  \\"\\"\\"\\"\\"\\"
  userId: Int!

  \\"\\"\\"\\"\\"\\"
  userName: String!

  \\"\\"\\"\\"\\"\\"
  text: String!

  \\"\\"\\"\\"\\"\\"
  loc: LocationInput

  \\"\\"\\"\\"\\"\\"
  tweetType: TweetType

  \\"\\"\\"\\"\\"\\"
  language: String
}

\\"\\"\\"\\"\\"\\"
input LocationInput {
  \\"\\"\\"\\"\\"\\"
  latitude: Float!

  \\"\\"\\"\\"\\"\\"
  longitude: Float!
}

\\"\\"\\"\\"\\"\\"
enum TweetType {
  \\"\\"\\"\\"\\"\\"
  TWEET

  \\"\\"\\"\\"\\"\\"
  RETWEET

  \\"\\"\\"\\"\\"\\"
  DM

  \\"\\"\\"\\"\\"\\"
  REPLY
}

\\"\\"\\"\\"\\"\\"
type TweetSearchResult {
  \\"\\"\\"\\"\\"\\"
  tweets: [Tweet]
}

\\"\\"\\"\\"\\"\\"
type Tweet {
  \\"\\"\\"\\"\\"\\"
  userId: Int!

  \\"\\"\\"\\"\\"\\"
  userName: String!

  \\"\\"\\"\\"\\"\\"
  text: String!

  \\"\\"\\"\\"\\"\\"
  loc: Location

  \\"\\"\\"\\"\\"\\"
  tweetType: TweetType

  \\"\\"\\"\\"\\"\\"
  language: String
}

\\"\\"\\"\\"\\"\\"
type Location {
  \\"\\"\\"\\"\\"\\"
  latitude: Float!

  \\"\\"\\"\\"\\"\\"
  longitude: Float!
}
"
`;
